api/src/exam-environment/routes/exam-environment.test.ts

```diff
@@ -0,0 +1,601 @@
+            examEnvironmentAuthorizationToken
+          )
+          .send(body);
+
+        expect(res).toMatchObject({
+          status: 400,
+          body: {
+            code: 'FCC_EINVAL_EXAM_ENVIRONMENT_EXAM_ATTEMPT'
+          }
+        });
+
+        // Database should mark attempt as `needsRetake`
+        const updatedAttempt =
+          await fastifyTestInstance.prisma.envExamAttempt.findUnique({
+            where: { id: attempt.id }
+          });
+        expect(updatedAttempt).toMatchObject({
+          needsRetake: true
+        });
+      });
+
+      it('should return 200 if request is valid, and update attempt in database', async () => {
+        const attempt = await fastifyTestInstance.prisma.envExamAttempt.create({
+          data: {
+            userId: defaultUserId,
+            examId: mock.examId,
+            generatedExamId: mock.generatedExam.id,
+            startTimeInMS: Date.now(),
+            questionSets: [],
+            needsRetake: false
+          }
+        });
+
+        const body: Static<typeof examEnvironmentPostExamAttempt.body> = {
+          attempt: mock.examAttemptSansSubmissionTimeInMS
+        };
+
+        const res = await superPost('/exam-environment/exam/attempt')
+          .set(
+            'exam-environment-authorization-token',
+            examEnvironmentAuthorizationToken
+          )
+          .send(body);
+
+        expect(res.status).toBe(200);
+
+        // Database should update attempt
+        const updatedAttempt =
+          await fastifyTestInstance.prisma.envExamAttempt.findUnique({
+            where: { id: attempt.id }
+          });
+
+        expect(updatedAttempt).toMatchObject(body.attempt);
+      });
+    });
+
+    describe('POST /exam-environment/generated-exam', () => {
+      afterEach(async () => {
+        await fastifyTestInstance.prisma.envExamAttempt.deleteMany();
+        await mock.seedEnvExam();
+      });
+
+      it('should return an error if the given exam id is invalid', async () => {
+        const body: Static<typeof examEnvironmentPostExamGeneratedExam.body> = {
+          examId: mock.oid()
+        };
+        const res = await superPost('/exam-environment/exam/generated-exam')
+          .send(body)
+          .set(
+            'exam-environment-authorization-token',
+            examEnvironmentAuthorizationToken
+          );
+
+        expect(res).toMatchObject({
+          status: 404,
+          body: {
+            code: 'FCC_ENOENT_EXAM_ENVIRONMENT_MISSING_EXAM'
+          }
+        });
+      });
+
+      xit('should return an error if the exam prerequisites are not met', async () => {
+        // TODO: Waiting on prerequisites
+      });
+
+      it('should return an error if the exam has been attempted in the last 24 hours', async () => {
+        const recentExamAttempt = {
+          ...mock.examAttempt,
+          // Set start time such that exam has expired, but 0 hours have passed
```

Comment:
Zero hours since the expiration? If so, how about "Set start time such that the exam has just expired"?

Reply:
Well, same thing. I suppose the code just needs to fulfill:
- Set start time such that the exam has expired, but less than 24 hours have passed.

Reply:
It might be me being weird, but I keep thinking "passed since what?". Also "0 hours" implies that the hours are important somehow. They aren't: all we care about is that it's expired, but inside the cooldown.

Tbh, I still think the exams need a terminology and "cooldown" needs to be in it.

Anyways, that's out of scope. For this PR I recommend "Set start time such that the exam has just expired"